### PR TITLE
Enhancement - Enabled external links on AppDetails page

### PR DESCRIPTION
**Pull request by sohel-rp**
# Description

Enabled users to configure quick links (external links) to the tools they use from the global configuration page for specific debugging/monitoring purposes. And those links can be accessed for specific app/pod/container from the AppDetails page.

**UX**
- [Global Configuration](https://www.figma.com/proto/z5oh20tbYNDZkDQGrSYgol/Global-Configurations?node-id=501%3A1272&scaling=min-zoom&page-id=501%3A1272&starting-point-node-id=520%3A1494&show-proto-sidebar=1)
- [App Details](https://www.figma.com/proto/j7cqlBdhQ47Bo10TLfamGl/Helm-Apps?node-id=603%3A1486&scaling=min-zoom&page-id=545%3A827&starting-point-node-id=603%3A1486&show-proto-sidebar=1)

Fixes # https://github.com/devtron-labs/devtron/issues/1422

# Screenshots

**Add external links from global configuration -** 

**App & pod level external links -**

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

All tests have been performed manually.

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/externalLinks/externalLinks.scss [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    .external-link__divider {
+        height: 1px;
+        width: 100%;
+        background-color: #EDF1F5;
+    }
+
+    .external-links__header {
+        height: 40px;
```
Comment (vivek-devtron) [inline src/components/externalLinks/externalLinks.scss:186]: can we minimize CSS here?

File: src/components/common/helpers/Helpers.tsx [hunk 2]
```diff
@@ -880,3 +880,21 @@ export function useSearchString(): UseSearchString {
 
     return { queryParams, searchParams }
 }
+
+export const sortOptionsByLabel = (optionA, optionB) => {
```
Comment (vivek-devtron) [inline src/components/common/helpers/Helpers.tsx:884]: something similar is already there in utils may be. just check and see if we can utilize that.

File: src/components/externalLinks/ExternalLinks.component.tsx [hunk 3]
```diff
@@ -0,0 +1,1030 @@
...
+                        onChange={(selected) => onMonitoringToolSelection(index, selected)}
+                        components={{
+                            IndicatorSeparator: null,
+                            ClearIndicator: null,
+                            Option: OptionWithIcon,
+                            ValueContainer: ValueContainerWithIcon,
+                        }}
+                        styles={{
+                            ...multiSelectStyles,
+                            container: (base, state) => ({
```
Comment (vivek-devtron) [inline src/components/externalLinks/ExternalLinks.component.tsx:364]: can we pull out common properties so that we can reuse it.

File: src/components/common/helpers/Helpers.tsx (hunk 2 above)
Comment (sohel-rp) [reply src/components/common/helpers/Helpers.tsx:884]: Yes. Tried using that util but there's some type mis-match issue so as discussed going ahead with these functions.

**Review by vivek-devtron — APPROVED** [review]

File: src/components/externalLinks/ExternalLinks.component.tsx (hunk 3 above)
Comment (sohel-rp) [reply src/components/externalLinks/ExternalLinks.component.tsx:374]: Done. Moved common styles to utils.

File: src/components/externalLinks/externalLinks.scss (hunk 1 above)
Comment (sohel-rp) [reply src/components/externalLinks/externalLinks.scss:198]: Yes. Tried to keep the CSS to a minimum.

File: src/components/EnvironmentOverride/ConfigMapOverrides.tsx [hunk 4]
```diff
@@ -474,7 +474,7 @@ const OverrideConfigMapForm: React.FC<ConfigMapProps> = memo(function OverrideCo
                 subPath={subPath}
                 filePermission={filePermission}
             />}
-        {state.dialog && <ConfirmationDialog className="confirmation-dialog__body--w-360">
+        {state.dialog && <ConfirmationDialog className="confirmation-dialog__body--w-400">
```
Comment (manish-agrawal-ai) [inline src/components/EnvironmentOverride/ConfigMapOverrides.tsx:477]: is this change intentional?

File: src/util/DeleteComponent.tsx [hunk 5]
```diff
@@ -50,7 +50,7 @@ function DeleteComponent({
 
     const renderCannotDeleteDialogModal = () => {
         return (
-            <ConfirmationDialog className="confirmation-dialog__body--w-360">
+            <ConfirmationDialog className="confirmation-dialog__body--w-400">
```
Comment (manish-agrawal-ai) [inline src/util/DeleteComponent.tsx:53]: is this change intentional?

Comment (sohel-rp) [reply src/util/DeleteComponent.tsx:53]: Yes. Made the change as part of the UX suggestions.

File: src/components/EnvironmentOverride/ConfigMapOverrides.tsx (hunk 4 above)
Comment (sohel-rp) [reply src/components/EnvironmentOverride/ConfigMapOverrides.tsx:477]: Yes. Made the change as part of the UX suggestions.

File: src/components/common/dialogs/DeleteDialog.tsx [hunk 6]
```diff
@@ -8,11 +9,12 @@ export interface DeleteDialogProps {
     closeDelete: () => void;
     delete: () => any;
     deletePrefix?: string
+    apiCallInProgress?: boolean
 }
 
 export const DeleteDialog: React.FC<DeleteDialogProps> & { Description?: React.FC<any> } = function (props) {
 
-    return <ConfirmationDialog className="confirmation-dialog__body--w-360">
+    return <ConfirmationDialog className="confirmation-dialog__body--w-400">
```
Comment (manish-agrawal-ai) [inline src/components/common/dialogs/DeleteDialog.tsx:17]: is this change intentional?

Comment (sohel-rp) [reply src/components/common/dialogs/DeleteDialog.tsx:17]: Yes

File: src/components/v2/values/common/ChartValuesSelectors.tsx [hunk 7]
```diff
@@ -184,10 +184,7 @@ export const ChartRepoSelector = ({
             <components.MenuList {...props}>
                 {props.children}
                 <div
-                    className="flex react-select__bottom bcn-0 pt-10 pb-2"
-                    style={{
-                        bottom: '-4px',
-                    }}
+                    className="flex react-select__bottom bcn-0"
```
Comment (manish-agrawal-ai) [inline src/components/v2/values/common/ChartValuesSelectors.tsx:187]: is this change intentional? (not related to this story though)

File: src/components/v2/values/common/ChartValuesSelectors.tsx [hunk 8]
```diff
@@ -223,6 +220,11 @@ export const ChartRepoSelector = ({
                             MenuList: customMenuListItem,
                         }}
                         styles={{
+                            menuList: (base) => ({
```
Comment (manish-agrawal-ai) [inline src/components/v2/values/common/ChartValuesSelectors.tsx:223]: is this change intentional? (not related to this story though)

File: src/components/v2/appDetails/ea/EAAppDetail.component.tsx [hunk 9]
```diff
@@ -77,13 +86,42 @@ function ExternalAppDetail({appId, appName}) {
...
+                                            value: tool.id,
+                                            icon: tool.icon,
+                                        }))
+                                        .sort(sortOptionsByValue) || [],
+                            })
+                            setIsLoading(false)
+                            isAPICallInProgress = false
+                        })
+                        .catch((e) => {
+                            setExternalLinksAndTools(externalLinksAndTools)
```
Comment (manish-agrawal-ai) [inline src/components/v2/appDetails/ea/EAAppDetail.component.tsx:116]: no error handling?

File: src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx [hunk 10]
```diff
@@ -139,71 +181,89 @@ function NodeComponent({handleFocusTabs}: {handleFocusTabs: () => void}) {
...
                             </div>
+                            {node.kind === NodeType.Pod && podLevelExternalLinks.length > 0 && (
+                                <NodeLevelExternalLinks
+                                    helmAppDetails={appDetails}
+                                    nodeLevelExternalLinks={podLevelExternalLinks}
+                                    podName={node.name}
+                                />
+                            )}
+                            {node.kind === NodeType.Containers && containerLevelExternalLinks.length > 0 && (
+                                <NodeLevelExternalLinks
```
Comment (manish-agrawal-ai) [inline src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:259]: Please code is generic way so that if external link comes on any level then integration is easy. for eg : on dpeloyment level

File: src/components/v2/values/common/ChartValuesSelectors.tsx (hunk 7 above)
Comment (sohel-rp) [reply src/components/v2/values/common/ChartValuesSelectors.tsx:187]: Yes. There was one visual/UI issue with the custom sticky bottom in react-select. There's some gap between the sticky bottom and the actual dropdown bottom causing text to be visible while scrolling through options. It's discovered as part of this story so fixed it.

The fix includes the removal of extra padding.

File: src/components/v2/values/common/ChartValuesSelectors.tsx (hunk 8 above)
Comment (sohel-rp) [reply src/components/v2/values/common/ChartValuesSelectors.tsx:223]: Yes.

File: src/components/common/modals/VisibleModal.tsx [hunk 11]
```diff
@@ -11,8 +11,12 @@ export class VisibleModal extends React.Component<{ className: string, noBackgro
     }
 
     escFunction(event) {
-        if (event.keyCode === 27 && this.props.close) {
-            this.props.close(event);
+        if (event.keyCode === 27 || event.key === "Escape") {
+            if (this.props.onEscape) {
+                this.props.onEscape(event);
+            } else if (this.props.close) {
```
[the remark this answers is not in the document]
Comment (sohel-rp) [reply src/components/common/modals/VisibleModal.tsx:17]: Yes. This will work only when either `onEscape` or `close` prop is passed.

This was added because in the current implementation when we use the close prop, it's closing the `VisibalModal` whenever we click on the modal. It's because `close` function is called on the click event of the `VisibleModal` as well.

So didn't touch the actual implementation however provided this extra prop to be used only on the escape button event.

File: src/components/v2/appDetails/ea/EAAppDetail.component.tsx (hunk 9 above)
Comment (sohel-rp) [reply src/components/v2/appDetails/ea/EAAppDetail.component.tsx:116]: As discussed, for error handling setting the old list instead of halting the page load or showing any error for this particular section on the page.

File: src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx (hunk 10 above)
Comment (sohel-rp) [reply src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:259]: Sure. will try to refactor this to be more generic.

Comment (sohel-rp) [reply src/components/v2/appDetails/k8Resource/nodeType/Node.component.tsx:259]: As discussed, keeping it only for 2 kinds for now i.e. `Pod` & `Container`. Will make it more generic to support other kinds as part of enhancement.

**Review by vivek-devtron — APPROVED** [review]
